File: src/domain/shifts.ts
import { LIMITS } from './limits';
import { MINUTE, minutesBetween } from './time';
import { Block, RestPeriod, blockMinutes, buildBlocks, findRestPeriods } from './timeline';
import type { ActivityEntry, RestStatus } from './types';

/** Смена, посчитанная по записям: от конца одного суточного отдыха до начала следующего. */
export interface DerivedShift {
  id: string;
  start: number;
  /** Начало отдыха, завершившего смену; null — смена идёт. */
  end: number | null;
  blocks: Block[];
  driveMinutes: number;
  workMinutes: number;
  poaMinutes: number;
  breakMinutes: number;
  /** В смене был отдых ≥ 3 ч — первая часть раздельного суточного отдыха 3 + 9. */
  splitFirstPart: boolean;
  continuousDriveAtEnd: number;
  restAfter: RestPeriod | null;
}

export interface Timeline {
  blocks: Block[];
  rests: RestPeriod[];
  shifts: DerivedShift[];
  /** Смена, которая идёт сейчас (последняя без завершающего отдыха). */
  current: DerivedShift | null;
}

/**
 * Отдых, который завершает смену: суточный (≥ 9 ч) или недельный, а также
 * идущий отдых, которым водитель завершил день.
 */
export const isShiftEndingRest = (p: RestPeriod): boolean =>
  p.restMinutes >= LIMITS.dailyRestReduced || (p.open && p.dayEnd);
export const isWeeklyRest = (p: RestPeriod): boolean => p.restMinutes >= LIMITS.weeklyRestReduced;

export function analyzeTimeline(entries: ActivityEntry[], now: number): Timeline {
  const blocks = buildBlocks(entries, now);
  const rests = findRestPeriods(blocks);
  const endingByFirstBlock = new Map<number, RestPeriod>();
  for (const p of rests) if (isShiftEndingRest(p)) endingByFirstBlock.set(p.firstBlock, p);

  const shifts: DerivedShift[] = [];
  let segment: Block[] = [];
  let i = 0;
  while (i < blocks.length) {
    const ending = endingByFirstBlock.get(i);
    if (ending) {
      const shift = makeShift(segment, ending);
      if (shift) shifts.push(shift);
      segment = [];
      i = ending.lastBlock + 1;
      continue;
    }
    segment.push(blocks[i]);
    i++;
  }
  const tail = makeShift(segment, null);
  if (tail) shifts.push(tail);

  const last = shifts[shifts.length - 1];
  return { blocks, rests, shifts, current: last && last.end === null ? last : null };
}

function makeShift(segment: Block[], restAfter: RestPeriod | null): DerivedShift | null {
  // Короткий отдых в начале сегмента — хвост данных, не часть смены.
  const firstWork = segment.findIndex((b) => b.activity !== 'REST');
  if (firstWork < 0) return null;
  const blocks = segment.slice(firstWork);

  let driveMinutes = 0;
  let workMinutes = 0;
  let poaMinutes = 0;
  let breakMinutes = 0;
  let splitFirstPart = false;
  for (const b of blocks) {
    const m = blockMinutes(b);
    if (b.activity === 'DRIVE') driveMinutes += m;
    else if (b.activity === 'WORK') workMinutes += m;
    else if (b.activity === 'POA') poaMinutes += m;
    else {
      breakMinutes += m;
      if (m >= LIMITS.dailyRestSplitFirst) splitFirstPart = true;
    }
  }

  const start = blocks[0].start;
  return {
    id: `auto-${start}`,
    start,
    end: restAfter ? restAfter.start : null,
    blocks,
    driveMinutes,
    workMinutes,
    poaMinutes,
    breakMinutes,
    splitFirstPart,
    continuousDriveAtEnd: computeBreakState(blocks).continuousDriveMinutes,
    restAfter,
  };
}

export interface BreakState {
  /** Вождение с последнего засчитанного перерыва. */
  continuousDriveMinutes: number;
  /** Взята первая часть раздельного перерыва (≥ 15 мин), вторая (≥ 30) ещё нет. */
  firstPartTaken: boolean;
  firstPart: { start: number; minutes: number } | null;
  /** Текущий отдых, если водитель отдыхает сейчас. */
  currentRest: { start: number; minutes: number; firstPartBefore: boolean } | null;
}

/**
 * Перерыв по ст. 7: 45 мин подряд или 15 + 30 именно в таком порядке.
 * Каждый блок отдыха обрабатывается один раз — текущий отдых тоже, поэтому
 * один отдых не может засчитаться сразу и первой, и второй частью.
 */
export function computeBreakState(blocks: Block[]): BreakState {
  let continuous = 0;
  let firstPart: BreakState['firstPart'] = null;
  let currentRest: BreakState['currentRest'] = null;

  for (const b of blocks) {
    const m = blockMinutes(b);
    if (b.activity === 'DRIVE') {
      continuous += m;
      continue;
    }
    if (b.activity !== 'REST') continue;

    if (b.open) currentRest = { start: b.start, minutes: m, firstPartBefore: firstPart !== null };
    if (m >= LIMITS.breakFull || (firstPart && m >= LIMITS.breakSplitSecond)) {
      continuous = 0;
      firstPart = null;
    } else if (!firstPart && m >= LIMITS.breakSplitFirst) {
      firstPart = { start: b.start, minutes: m };
    }
  }

  return {
    continuousDriveMinutes: continuous,
    firstPartTaken: firstPart !== null,
    firstPart,
    currentRest,
  };
}

/** Окно, в котором должен пройти суточный отдых: 24 ч от начала смены, экипаж — 30 ч. */
export const shiftWindowMinutes = (team: boolean): number =>
  team ? LIMITS.shiftWindowTeam : LIMITS.shiftWindowSolo;

/**
 * Часть суточного отдыха внутри окна от начала смены. По ст. 8(2) статус
 * отдыха определяет именно она: смена 14 ч и отдых 12 ч — сокращённый отдых,
 * в окне только 10 ч.
 */
export function restInWindow(spanMinutes: number, restMinutes: number, team: boolean): number {
  return Math.max(0, Math.min(restMinutes, shiftWindowMinutes(team) - spanMinutes));
}

/** То же для смены из записей: прерывания отдыха на пароме в окно не входят. */
export function shiftRestInWindow(blocks: Block[], shift: DerivedShift, team: boolean): number {
  const rest = shift.restAfter;
  if (!rest) return 0;
  const windowEnd = shift.start + shiftWindowMinutes(team) * MINUTE;
  let minutes = 0;
  for (let k = rest.firstBlock; k <= rest.lastBlock; k++) {
    const b = blocks[k];
    if (b.activity === 'REST') minutes += minutesBetween(b.start, Math.min(b.end, windowEnd));
  }
  return minutes;
}

/** Статус суточного отдыха по его длительности (в окне — см. restInWindow). */
export function dailyRestStatus(minutes: number, split: boolean): RestStatus {
  if (minutes >= LIMITS.dailyRestRegular) return 'full';
  if (split && minutes >= LIMITS.dailyRestSplitSecond) return 'full';
  if (minutes >= LIMITS.dailyRestReduced) return 'reduced';
  return 'insufficient';
}

/** Статус недельного отдыха по его длительности. */
export function weeklyRestStatus(minutes: number): RestStatus {
  if (minutes >= LIMITS.weeklyRestRegular) return 'full';
  if (minutes >= LIMITS.weeklyRestReduced) return 'reduced';
  return 'insufficient';
}

File: packages/tacho_engine/lib/src/shifts.dart
import 'package:meta/meta.dart';
import 'package:tacho_engine/src/activity_period.dart';
import 'package:tacho_engine/src/driver_mode.dart';
import 'package:tacho_engine/src/eu_limits.dart';
import 'package:tacho_engine/src/time.dart';
import 'package:tacho_engine/src/timeline.dart';

/// Состав экипажа: от него зависит окно суточного отдыха (ст. 8(5)).
enum CrewMode {
  /// Один водитель: 11 ч (9 ч) отдыха в пределах 24 ч.
  solo,

  /// Экипаж из двух водителей: 9 ч отдыха в пределах 30 ч.
  team,
}

/// Оценка отдыха по его длительности.
enum RestStatus { full, reduced, insufficient }

/// Смена, посчитанная по записям: от конца одного суточного отдыха до
/// начала следующего.
@immutable
class Shift {
  const new({
    required this.start,
    required this.end,
    required this.blocks,
    required this.driving,
    required this.otherWork,
    required this.availability,
    required this.breaks,
    required this.splitFirstPart,
    required this.continuousDrivingAtEnd,
    required this.restAfter,
  });

  final DateTime start;

  /// Начало отдыха, завершившего смену; null — смена идёт.
  final DateTime? end;
  final List<Block> blocks;
  final Duration driving;
  final Duration otherWork;
  final Duration availability;
  final Duration breaks;

  /// В смене был отдых ≥ 3 ч — первая часть раздельного суточного отдыха
  /// 3 + 9 (ст. 8(2)).
  final bool splitFirstPart;
  final Duration continuousDrivingAtEnd;
  final RestPeriod? restAfter;

  bool get isOngoing => end == null;

  @override
  String toString() =>
      'Shift($start → ${end ?? 'now'}, '
      'вождение ${driving.inMinutes} мин)';
}

/// Журнал, разобранный на блоки, периоды отдыха и смены.
@immutable
class Timeline {
  const new({
    required this.blocks,
    required this.rests,
    required this.shifts,
    required this.current,
  });

  final List<Block> blocks;
  final List<RestPeriod> rests;
  final List<Shift> shifts;

  /// Смена, которая идёт сейчас (последняя без завершающего отдыха).
  final Shift? current;
}

/// Разбирает журнал на смены. Смену завершает отдых ≥ 9 ч (или идущий
/// отдых, объявленный концом дня); более короткий отдых — перерыв внутри
/// смены.
Timeline analyzeTimeline(Iterable<ActivityPeriod> periods, DateTime now) {
  final blocks = buildBlocks(periods, now);
  final rests = findRestPeriods(blocks);
  final endingByFirstBlock = {
    for (final p in rests)
      if (p.endsShift) p.firstBlock: p,
  };

  final shifts = <Shift>[];
  var segment = <Block>[];
  var i = 0;
  while (i < blocks.length) {
    final ending = endingByFirstBlock[i];
    if (ending != null) {
      final shift = _makeShift(segment, ending);
      if (shift != null) shifts.add(shift);
      segment = [];
      i = ending.lastBlock + 1;
      continue;
    }
    segment.add(blocks[i]);
    i++;
  }
  final tail = _makeShift(segment, null);
  if (tail != null) shifts.add(tail);

  final last = shifts.isEmpty ? null : shifts.last;
  return Timeline(
    blocks: blocks,
    rests: rests,
    shifts: shifts,
    current: last != null && last.isOngoing ? last : null,
  );
}

Shift? _makeShift(List<Block> segment, RestPeriod? restAfter) {
  // Короткий отдых в начале сегмента — хвост данных, не часть смены.
  final firstWork = segment.indexWhere((b) => !b.mode.isRest);
  if (firstWork < 0) return null;
  final blocks = segment.sublist(firstWork);

  var driving = Duration.zero;
  var otherWork = Duration.zero;
  var availability = Duration.zero;
  var breaks = Duration.zero;
  var splitFirstPart = false;
  for (final b in blocks) {
    final d = b.duration;
    switch (b.mode) {
      case DriverMode.driving:
        driving += d;
      case DriverMode.otherWork:
        otherWork += d;
      case DriverMode.availability:
        availability += d;
      case DriverMode.rest:
        breaks += d;
        if (d >= EuLimits.dailyRestSplitFirst) splitFirstPart = true;
    }
  }

  return Shift(
    start: blocks.first.start,
    end: restAfter?.start,
    blocks: blocks,
    driving: driving,
    otherWork: otherWork,
    availability: availability,
    breaks: breaks,
    splitFirstPart: splitFirstPart,
    continuousDrivingAtEnd: computeBreakState(blocks).continuousDriving,
    restAfter: restAfter,
  );
}

/// Часть раздельного перерыва или идущий перерыв.
@immutable
class BreakPart {
  const new({required this.start, required this.duration});

  final DateTime start;
  final Duration duration;

  @override
  bool operator ==(Object other) =>
      other is BreakPart && other.start == start && other.duration == duration;

  @override
  int get hashCode => Object.hash(start, duration);
}

/// Перерыв, который идёт сейчас.
@immutable
class CurrentRest {
  const new({
    required this.start,
    required this.duration,
    required this.firstPartBefore,
  });

  final DateTime start;
  final Duration duration;

  /// До этого отдыха взята первая часть раздельного перерыва: ему хватит
  /// 30 мин.
  final bool firstPartBefore;
}

@immutable
class BreakState {
  const new({
    required this.continuousDriving,
    required this.firstPart,
    required this.currentRest,
  });

  /// Вождение с последнего засчитанного перерыва.
  final Duration continuousDriving;

  /// Первая часть раздельного перерыва (≥ 15 мин), если вторая (≥ 30) ещё
  /// не взята.
  final BreakPart? firstPart;

  /// Текущий отдых, если водитель отдыхает сейчас.
  final CurrentRest? currentRest;

  bool get firstPartTaken => firstPart != null;
}

/// Перерыв по ст. 7: 45 мин подряд или 15 + 30 именно в таком порядке.
/// Каждый блок отдыха обрабатывается один раз — текущий отдых тоже, поэтому
/// один отдых не может засчитаться сразу и первой, и второй частью.
BreakState computeBreakState(List<Block> blocks) {
  var continuous = Duration.zero;
  BreakPart? firstPart;
  CurrentRest? currentRest;

  for (final b in blocks) {
    final d = b.duration;
    if (b.mode == DriverMode.driving) {
      continuous += d;
      continue;
    }
    if (!b.mode.isRest) continue;

    if (b.open) {
      currentRest = CurrentRest(
        start: b.start,
        duration: d,
        firstPartBefore: firstPart != null,
      );
    }
    if (d >= EuLimits.breakFull ||
        (firstPart != null && d >= EuLimits.breakSplitSecond)) {
      continuous = Duration.zero;
      firstPart = null;
    } else if (firstPart == null && d >= EuLimits.breakSplitFirst) {
      firstPart = BreakPart(start: b.start, duration: d);
    }
  }

  return BreakState(
    continuousDriving: continuous,
    firstPart: firstPart,
    currentRest: currentRest,
  );
}

/// Окно, в котором должен пройти суточный отдых: 24 ч от начала смены,
/// экипаж — 30 ч.
Duration workdayWindow(CrewMode crew) => switch (crew) {
  CrewMode.solo => EuLimits.workdayWindow,
  CrewMode.team => EuLimits.teamWorkdayWindow,
};

/// Часть суточного отдыха внутри окна от начала смены. По ст. 8(2) статус
/// отдыха определяет именно она: смена 14 ч и отдых 12 ч — сокращённый
/// отдых, в окне только 10 ч.
Duration restInWindow(Duration span, Duration rest, CrewMode crew) =>
    clampToZero(shorter(rest, workdayWindow(crew) - span));

/// То же для смены из записей: прерывания отдыха на пароме в окно не входят.
Duration shiftRestInWindow(List<Block> blocks, Shift shift, CrewMode crew) {
  final rest = shift.restAfter;
  if (rest == null) return Duration.zero;
  final windowEnd = shift.start.add(workdayWindow(crew));
  var total = Duration.zero;
  for (var k = rest.firstBlock; k <= rest.lastBlock; k++) {
    final b = blocks[k];
    if (b.mode.isRest) {
      total += durationBetween(b.start, earlier(b.end, windowEnd));
    }
  }
  return total;
}

/// Статус суточного отдыха по его длительности (в окне — см.
/// [restInWindow]).
RestStatus dailyRestStatus(Duration rest, {required bool split}) {
  if (rest >= EuLimits.dailyRestRegular) return RestStatus.full;
  if (split && rest >= EuLimits.dailyRestSplitSecond) return RestStatus.full;
  if (rest >= EuLimits.dailyRestReduced) return RestStatus.reduced;
  return RestStatus.insufficient;
}

/// Статус недельного отдыха по его длительности.
RestStatus weeklyRestStatus(Duration rest) {
  if (rest >= EuLimits.weeklyRestRegular) return RestStatus.full;
  if (rest >= EuLimits.weeklyRestReduced) return RestStatus.reduced;
  return RestStatus.insufficient;
}
